vault/aws/iam.go

```diff
@@ -0,0 +1,125 @@
+package aws
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/json"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials/stscreds"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/sts"
+	"github.com/hashicorp/vault/logical"
+	"github.com/pkg/errors"
+	"github.com/securityscorecard/vault-vouch/vault"
+)
+
+const awsAuthUrl = "/v1/auth/aws/login"
+
+var httpClient = http.DefaultClient
+
+type generator struct {
+	sts       *sts.STS
+	vaultAddr string
+}
+
+// DefaultGenerator returns a vault.Generator implementation that uses the autoconfigured
+// AWS credentials in the current environment
+func DefaultGenerator(vaultAddress string) vault.Generator {
+	sess := session.Must(session.NewSession())
+	s := sts.New(sess)
+	return &generator{
+		sts:       s,
+		vaultAddr: vaultAddress,
+	}
+}
+
+// AssumeRoleGenerator returns a vault.Generator implementation that uses the provided
+// AWS role before generating the login payload for Vault
+func AssumeRoleGenerator(vaultAddress string, roleArn string) vault.Generator {
+	sess := session.Must(session.NewSession())
+	creds := stscreds.NewCredentials(sess, roleArn)
+	return &generator{
+		sts:       sts.New(sess, &aws.Config{Credentials: creds}),
+		vaultAddr: vaultAddress,
+	}
+}
+
+// WrappedToken returns a token for the role provided with the wrapTTL provided
+// WARNING: If wrapTTL is a zero length duration we provide an unwrapped token
+func (a generator) WrappedToken(role string, wrapTTL time.Duration) (*string, error) {
```

Comment:
I think usually it's preferred to return `"", err` (meaning the method returns `(string, error)`) unless there's a special reason to return a `*string`.

Thinking, as a potential user of a library, that it's not as nice having to work with pointers to primitive types.

Reply:
Also, any reason not to define the method on a pointer receiver? `func (a *generator) ...`

Reply:
Not modifying any of the struct members so no pointer receiver.

Reply:
Point well taken around `*string`, that's awkward. Will fix.